app/uk/gov/hmrc/customs/notification/controllers/HeaderValidator.scala

```diff
@@ -43,92 +45,98 @@ trait HeaderValidator {
     override def parser: BodyParser[AnyContent] = controllerComponents.parsers.defaultBodyParser
     def invokeBlock[A](request: Request[A], block: Request[A] => Future[Result]): Future[Result] = {
       implicit val headers: Headers = request.headers
-      val logMessage = "Received notification"
-      notificationLogger.debugWithHeaders(logMessage, headers.headers)
-
-      if (!hasAccept) {
-        Future.successful(ErrorAcceptHeaderInvalid.XmlResult)
-      } else if (!hasContentType) {
-        Future.successful(ErrorContentTypeHeaderInvalid.XmlResult)
-      } else if (missingClientId) {
-        Future.successful(ErrorCdsClientIdMissing.XmlResult)
-      } else if (!hasValidClientId) {
-        Future.successful(ErrorCdsClientIdInvalid.XmlResult)
-      } else if (missingConversationId) {
-        Future.successful(ErrorConversationIdMissing.XmlResult)
-      } else if (!hasValidConversationId) {
-        Future.successful(ErrorConversationIdInvalid.XmlResult)
-      } else if (!hasAuth(maybeBasicAuthToken)) {
-        Future.successful(ErrorUnauthorized.XmlResult)
-      } else if(!correlationIdIsValidIfPresent) {
-        Future.successful(ErrorGenericBadRequest.XmlResult)
+
+      notificationLogger.debugWithHeaders("Received notification", headers.headers)
+
+      def collect(validation: Headers => Option[String], logMessage: String, error: ErrorResponse)(implicit h: Headers) = {
+        val maybeString = validation(h)
+        maybeString match {
+          // Wrap the Error in an exception - should not be using Error https://stackoverflow.com/questions/912334/differences-between-exception-and-error
+          // https://stackoverflow.com/questions/17265022/what-is-a-boxed-error-in-scala
+          case None =>
+            notificationLogger.debugWithPrefixedHeaders(logMessage, headers.headers)
+            Future.failed(new Exception(error))
+          case Some(msg) =>
+            Future.successful(logMessage + "\n" + msg)
+        }
       }
-      else {
+
+      val logOutput: Future[String] = for {
+        a <- collect(hasAccept, "", ErrorAcceptHeaderInvalid)
```

Comment:
Rather than just a,b,c would prefer logAccumlatorA, logAccumlatorB, logAccumlatorC to give reader an idea of what it is